http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -0,0 +1,205 @@
+package org.broadinstitute.dsde.workbench.leonardo.monitor
+
+import cats.effect.{Async, Concurrent, ContextShift, IO, Timer}
+import org.broadinstitute.dsde.workbench.google2.{Event, GoogleSubscriber}
+import fs2.{Pipe, Stream}
+import io.circe.{Decoder, DecodingFailure, Encoder}
+import org.broadinstitute.dsde.workbench.leonardo.MachineConfig
+import org.broadinstitute.dsde.workbench.leonardo.JsonCodec._
+import org.broadinstitute.dsde.workbench.leonardo.db.{clusterQuery, followupQuery, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.model.google.{ClusterStatus, MachineType}
+import org.broadinstitute.dsde.workbench.leonardo.util.ClusterHelper
+import _root_.io.chrisdavenport.log4cats.Logger
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model.LeoException
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterStatus.Stopped
+import org.broadinstitute.dsde.workbench.model.WorkbenchException
+
+import scala.concurrent.ExecutionContext
+
+case class ClusterNotFoundException(
+  clusterId: Long,
+  override val message: String =
+    "Could not process ClusterTransitionFinishedMessage because it was not found in the database"
+) extends LeoException
+
+class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Logger: Concurrent](
+  subscriber: GoogleSubscriber[IO, LeoPubsubMessage],
+  clusterHelper: ClusterHelper,
+  dbRef: DbReference[IO]
+)(implicit executionContext: ExecutionContext)
+    extends LazyLogging {
+
+  def messageResponder(message: LeoPubsubMessage): IO[Unit] = {
+    val response = message match {
+      case msg @ StopUpdateMessage(_, _) =>
+        handleStopUpdateMessage(msg)
+      case msg @ ClusterTransitionFinishedMessage(_) =>
+        handleClusterTransitionFinished(msg)
+      case _ => IO.unit
+    }
+
+    //ensure we don't crash if the handler throws an exception
+    response.handleErrorWith(
+      e =>
+        IO(
+          logger
+            .error(s"Unable to process a message in received from pub/sub subscription in messageResponder: ${message}",
+                   e)
+        )
+    )
+
+    response
+  }
+
+  def messageHandler: Pipe[IO, Event[LeoPubsubMessage], Unit] = in => {
+    in.flatMap { event =>
+      for {
+        _ <- Stream
+          .eval(IO.pure(event.consumer.ack())) //we always ack first, as it could cause an endless loop of exceptions to do it after
+        _ <- Stream.eval(messageResponder(event.msg))
+      } yield ()
+    }
+  }
+
+  val process: Stream[IO, Unit] = (subscriber.messages through messageHandler).repeat
+
+  def handleStopUpdateMessage(message: StopUpdateMessage): IO[Unit] =
```

Comment:
I think all methods other than `process` can be marked as `private`?

Reply:
I did it for all but `messageResponder`, which is used heavily in the unit tests. It seems mockito is incapable of mocking an fs2 Stream via

```
queue.enqueue1(Event(message.asInstanceOf[LeoPubsubMessage], mock[AckReplyConsumer]))
    when {
      googleSubscriber.messages
    } thenReturn queue.dequeue
```